File: binary_tree/solve_problems_recursively/maximum_depth_of_a_binary_tree/maximum_depth_of_a_binary_tree.java
/**
 * Author: Nathan Bockisch
 * Date: October 13, 2021
 **/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int findMaxDepth(TreeNode cur, int prev_level) {
        if (cur == null) return prev_level;
        
        int level = prev_level + 1;
        int child_max = Math.max(findMaxDepth(cur.left, level), 
                        findMaxDepth(cur.right, level));
        return Math.max(level, child_max);
    }
    
    public int maxDepth(TreeNode root) {
        int level = 0;
        
        if (root != null) 
            level = findMaxDepth(root, level);
        
        return level;
    }
}

File: recursion_i/complexity_analysis/maximum_depth_of_a_binary_tree/maximum_depth_of_a_binary_tree.cpp
/**
 * Author: Nathan Bockisch
 * Date: October 22, 2021
 **/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countMaxDepth(TreeNode* cur, int count) {
        if (!cur) return count;
        
        count++;

        if (!cur->left && !cur->right) return count;
        
        return max(countMaxDepth(cur->left, count), countMaxDepth(cur->right, count));
    }
    
    int maxDepth(TreeNode* root) {
        return countMaxDepth(root, 0);
    }
};
